app/lib/test_applications.rb

```diff
@@ -1,56 +1,108 @@
 class TestApplications
   class NotEnoughCoursesError < RuntimeError; end
-  class ZeroCoursesPerApplicationError < RuntimeError; end
-
-  attr_reader :time
-
-  def generate_for_provider(provider:, courses_per_application:, count:, for_ratified_courses: false)
-    courses_to_apply_to = if for_ratified_courses
-                            GetCoursesRatifiedByProvider.call(provider: provider)
-                          else
-                            Course.current_cycle.includes(:course_options)
-                              .joins(:course_options).distinct.open_on_apply
-                              .where(provider: provider)
-                          end
-    1.upto(count).flat_map do
-      create_application(
-        recruitment_cycle_year: RecruitmentCycle.current_year,
-        states: [:awaiting_provider_decision] * courses_per_application,
-        courses_to_apply_to: courses_to_apply_to,
-      )
+  class ZeroCoursesPerApplicationError < RuntimeError
+    def message
+      'You cannot have zero courses per application'
+    end
+  end
+  class CourseAndStateNumbersDoNotMatchError < RuntimeError
+    def message
+      'The number of states and courses must be equal'
+    end
+  end
+  class OnlyOneCourseWhenApplyingAgainError < RuntimeError
+    def message
+      'You can only apply to one course when applying again'
     end
   end
 
+  attr_reader :time
+
   def create_application(
     recruitment_cycle_year:,
     states:,
     courses_to_apply_to:,
     apply_again: false,
     carry_over: false,
     course_full: false,
-    candidate: nil,
-    decline_by_default_at: 3.days.from_now
+    candidate: nil
+  )
+    courses = courses_to_apply_to(
+      states: states,
+      courses_to_choose_from: courses_to_apply_to,
+      course_full: course_full,
+    )
+    create_application_to_courses(
+      recruitment_cycle_year: recruitment_cycle_year,
+      states: states,
+      courses: courses,
+      apply_again: apply_again,
+      carry_over: carry_over,
+      candidate: candidate,
+    )
+  end
+
+private
+
+  def courses_to_apply_to(
```

Comment:
I did this refactor and then realised it didn't help with my error handling. It seems ok to leave it in though, as a first step towards breaking the huge method down?

Reply:
:+1: This is definitely better!

Reply:
Agree 👍